easy_phi/hwal.py

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import serial
+import threading
+from tornado import gen
+
+
+def lock(func):
+    """decorator to prevent concurrent execution of some function
+    This decorator should be used to wrap calls to devices to evade overlapping commands from different requests
+    For SCPI supporting devices, same can be achieved by using  SCPI commands *WAI, *OPC? or *STB?,
+    but lock will result in lower latency
+    """
+    def wrapper(self, *args, **kwargs):
+        if not hasattr(self, 'lock'):
+            self.lock = threading.Lock()
+            self.is_locked = lambda x: x.lock.locked()
+        self.lock.acquire()
+        retval = func(self, *args, **kwargs)
+        self.lock.release()
+        return retval
+    return wrapper
+
+
+class AbstractMeasurementModule(object):
+    """ Abstract class for representation of test measurement equipment
+    It should not be used directly but rather serve as a reference for
+    different module types subclasses
+    """
+
+    name = "Abstract module"
+
+    def __init__(self, device):
+        """ Initialize module object with pyudev.Device object """
+        self.device = device
+
+    @staticmethod
+    def is_instance(device):
+        """ Check if the device supported by this module
+        :param device: pyudev.Device instance
+        :return: boolean
+        """
+        return False
+
+    @lock
+    def scpi(self, command):
+        """Send SCPI command to device"""
+        return "OK"
```

Comment:
Hardcoded "OK" is used throughout this file. I'm cool if you say so. In Java it would be considered a bad coding practice